src/NovelAudioService.cpp

```diff
@@ -0,0 +1,367 @@
+    if (_channelMap.find(soundName) != _channelMap.end() && _channelMap[soundName] != -1)
+  {
+    Mix_SetPanning(_channelMap[soundName], leftChannelVolume, rightChannelVolume);
+  }
+  else
+  {
+    std::cerr << soundName << " is not allocated to a channel already!" << std::endl;
+  }
+}
+
+void NovelAudioService::loadMusic(std::string input) {
+  auto exists = _music.find(input);
+  if (exists != _music.end()) return;
+
+  Mix_Music* newMusic = Mix_LoadMUS(input.c_str());
+  if (newMusic != nullptr)
+  {
+    _music[input] = newMusic;
+  }
+  else
+  {
+    std::cerr << Mix_GetError() << std::endl;
+  }
+}
+
+void NovelAudioService::unloadMusic(std::string input) {
+  auto existingMusic = _music.find(input);
+  if (existingMusic == _music.end()) return;
+
+  Mix_FreeMusic(_music[input]);
+  _music.erase(existingMusic);
+}
+
+void NovelAudioService::playMusic() {
+  if (Mix_PausedMusic()) Mix_ResumeMusic();
+}
+
+void NovelAudioService::playMusic(std::string musicName) {
+  auto existingMusic = _music.find(musicName);
+  if (existingMusic == _music.end())
+  {
+    loadMusic(musicName);
+    existingMusic = _music.find(musicName);
+    if (existingMusic == _music.end()) return;
+  }
+  _channelMap[musicName] = -3;
+  Mix_PlayMusic(_music[musicName],-1);
+  _musicTime = SDL_GetPerformanceCounter();
+}
+
+void NovelAudioService::playMusic(std::string musicName, int loops) {
+
+  auto existingMusic = _music.find(musicName);
+  if (existingMusic == _music.end())
+  {
+    loadMusic(musicName);
+    existingMusic = _music.find(musicName);
+    if (existingMusic == _music.end()) return;
+  }
+  _channelMap[musicName] = -3;
+  Mix_PlayMusic(_music[musicName], loops-1);
+}
+
+void NovelAudioService::pauseMusic() {
+  if (!Mix_PausedMusic())
+  {
+    Mix_PauseMusic();
+    _musicPausedTime = SDL_GetPerformanceCounter();
+  }
+}
+
+void NovelAudioService::stopMusic() {
+  Mix_HaltMusic();
+  _musicPausedTime = SDL_GetPerformanceCounter();
+}
+
+void NovelAudioService::setMusicVolume(float value) {
+  Mix_VolumeMusic(convertToMixVolume(value));
+}
+
+void NovelAudioService::fadeMusicInOnce(std::string musicName, int ms) {
+  if (!Mix_PlayingMusic())
+  {
+    Mix_FadeInMusic(_music[musicName], 0, ms);
+  }
+  else
+  {
+    Mix_FadeOutMusic(1000);
+    switch(Mix_GetMusicType(_music[musicName]))
+    {
+      case MUS_OGG:
+      {
+        double pos = (double)((_musicPausedTime - _musicTime)*1000) / (double)SDL_GetPerformanceFrequency();
+        Mix_SetMusicPosition(pos);
+        Mix_FadeInMusic(_music[musicName], 0, ms);
+        break;
+      }
+      case MUS_MP3:
+      {
+        Mix_RewindMusic();
+        double pos = (double)((_musicPausedTime - _musicTime)*1000) / (double)SDL_GetPerformanceFrequency();
+        Mix_SetMusicPosition(pos);
+        Mix_FadeInMusic(_music[musicName], 0, ms);
+        break;
+      }
+      default:
+      {
+        Mix_FadeInMusic(_music[musicName], 0, ms);
```

Comment:
You know I'm running out of spells so I'll just leave this comment at the end of this block because boi do I feel like I'm in harry potter and the flying magic numbers right now.

Reply:
Fixed in latest commit 👍